bestofjs/bestofjs
# Improve search palette feature (debounce + refactoring)

## Goal

- Debounce calls to search functions in the Search palette to avoid blocking the UI while the user is typing
- Refactor the Search Palette feature, to split state and components, and introduce a library [Unstated Next](https://github.com/jamiebuilds/unstated-next) that is a thin wrapper around React Context API (unfortunately nobody knows it, but I like it very much)
- Print timing information in debug mode to measure the time it takes to perform searches
- Make search algorithm a bit faster by avoiding creating the same RegExp objects for each project the function loop through when ranking every project (about 2000 projects)
- Add basic tests for search functions

## How to test

Type very fast in the search input and ensure the search feature works!

Test command: `pnpm exec vitest src/lib/search-utils.test.ts`

![image](https://github.com/bestofjs/bestofjs/assets/5546996/25ebce4c-3386-46f1-8417-7ac63a722d10)

## Screenshot

![image](https://github.com/bestofjs/bestofjs/assets/5546996/ba128bf6-5afa-485c-8805-0971017d2747)
